### PR TITLE
PReLU shared axes support

### PR DESCRIPTION
**Pull request by PhilippKopp**
Started to feed the shared axes information to the layer. Extended the constructor to handle a size_t vector with the shared axes. Next step is to use that information in inference.

### CODE REVIEW

**Comment by Dobiasd** [comment]
You are right. For frugally-deep every tensor is converted to 3 dimensions. Empty dimensions are set to size `1`, somewhat like left-padding.

And your proposed solution makes sense to me. You can augment the list of shared axes either in the ctor or even at the beginning of `apply_impl`. Do whatever you feel is better. :slightly_smiling_face: 

One small style remark:
Instead of writing
```cpp
bool height_shared = std::find(shared_axes_.begin(), shared_axes_.end(), 1) != shared_axes_.end();
```

You can use:
```cpp
const bool height_shared = fplus::is_elem_of(1, shared_axes_);
```

In my opinion this is easier to read and more expressive.
Of course one does not need to have `fplus` names like this memorized. In most cases you know the type signature of the function you are looking for. In this particular case it would be `(a, [a]) -> bool`. Now you just have to type it in [here](http://www.editgym.com/fplus-api-search/) and the function will be presented to you. :books:

**Comment by PhilippKopp** [comment]
Thanks for the `is_elem_of` idea! Definitly agree :) looks way better.

Regarding the empty dimensions I'm still a bit lost.. Let's omit the batch dimension for the following and look at the last test case that fails. The input size in keras is a 2d `(None, 4)`. So the PReLU applied on that layer has to have the `None` dimension included in `shared_axes` which would be 1 (batch dimension is 0). 
In the PReLU layer `apply_impl` function the input tensor then has depth=4, height=1 (the dimension inserted) and width=42 (the None dimension). So now the 1 in `shared_axes` isn't correct anymore but should be set to 2. It could also be that the keras layer had a size of (1, 42, 4) and the user wanted to set dim 1 to shared. That wouldn't make a difference as the length in that dimension is 1 anyway, but I think we shouldn't just guess that the dimension was added by fdeep and set `shared_axes` to 2, because there can be cases where the guessing can be wrong.
I didn't find out where fdeep adds the additional dimension or where we could figure out that the keras layer was just 2 dimensional. Where ever that is we would then have to adjust the `shared_axes` list accordingly.

**Comment by Dobiasd** [comment]
The input and output tensors for the test data are converted in [arr_as_arr3 in convert_model.py](https://github.com/Dobiasd/frugally-deep/blob/42ad87b294ba0299908cc2d8704de13823c14f0c/keras_export/convert_model.py#L41-L53).

If there is no reliable way to determine what to add to `shared_axes`, we should try to somehow get the needed additional information from Python to C++. But if I understand correctly, in your example it is not just "guessing" (heuristics with possibly wrong results) but an unambiguous deduction. So if it is semantically the same to have a dimension as `1` or don't have it at all, we could treat it as the same situation. If I'm missing something, please let me know. I have a feeling that you are developing a better understanding of this dimension hassle than I have. :chart_with_upwards_trend:

**Comment by PhilippKopp** [comment]
Let me give a (hopefully) better example: 

Keras layer A:
PReLU with input size (None,4) and shared_axes=[1]

Keras layer B:
PReLU with input size (1,None,4) and shared_axes=[2]

Keras layer C:
PReLU with input size (42,4) and shared_axes=[1]

Keras layer D:
PReLU with input size (1,42,4) and shared_axes=[2]

Keras layer E (here not the dimension with len 42 is shared but dimension with len 1!):
PReLU with input size (1,42,4) and shared_axes=[1]

Now using an input for that layer with size (1,42,4) they all have the same shape in fdeep. For the 2D cases a third dimension is added. The problem is that although layers A-D should do the same thing we have different shared_axes and I don't see a way how to infer where an axes was added and how to distinguish cases A-D and case E.
One idea I had was to make case E impossible by deleting dimensions from the shared_axes list in python, if the dimension length is 1 which doesn't make a difference. Then we could check for that in c++ and shift accordingly.

In our use case the PReLUs with shared_axes are all 3D. So I am also fine with adding an assert that the PReLU input size is 3D, but I guess it would be nice to solve it more generally :)

**Comment by Dobiasd** [comment]
Ah OK, I think I understand the problem.

So the clean solution would be to not fill up dimensions with `1`s but instead explicitly also support shapes with 1 and 2 dimensions on C++ side?

**Comment by PhilippKopp** [comment]
I added the shifting that fixes shared_axes_ if the tensor was one or two dimensional in keras. It passes all the test cases now! Nevertheless I'm a bit confused. I had thought we would have to add some checks in python and delete axes from the list if their length is 1. Apparently that is not necessary. At least I wasn't able to create test cases that fail :)
Maybe you were right after all, that these situations can be treated in the same way!

**Comment by PhilippKopp** [comment]
I guess you had the better intuition that we could treat it the same way :) 
I tried like 20 different combinations of input shape and shared_axes and couldn't break it. 
I'm fine with merging!

**Comment by Dobiasd** [comment]
@PhilippKopp Just did some [minor clean-up](https://github.com/Dobiasd/frugally-deep/commit/c84d256f49e41b213d740999140dd7099a853491) on the new code that I wanted to share with you.

---

1) The constructor of `prelu_layer` had a default value for its `shared_axes` parameter:

```cpp
std::vector<std::size_t> shared_axes = std::vector<std::size_t>()
```

However when calling this ctor we had:

```cpp
if (!shared_axes.empty())
{  
    return std::make_shared<prelu_layer>(name, alpha, shared_axes);
}
return std::make_shared<prelu_layer>(name, alpha);
```

So I simplified it by removing the default value and always just calling:
```cpp
return std::make_shared<prelu_layer>(name, alpha, shared_axes);
```

which is the same but shorter/simpler.

---

2) In `prelu_layer::apply_impl` we had `alpha_->at(pos)`. I replaced it with `(*alpha_)[pos]` for performance reasons. [`std::vector::at`](https://en.cppreference.com/w/cpp/container/vector/at) does bounds checking. [`std::vector::operator[]`](https://en.cppreference.com/w/cpp/container/vector/operator_at) does not. Sure, bounds checking is nice, but I'd like to replace the CPU overhead with developer condifence. :wink: The other functions in frugally-deep that access vector elements (like e.g. `fdeep::tensor3::get`) also use the non-checked version.

---

3) The rest was mainly boring stuff like adding some newlines, `const` annotations etc.

**Comment by PhilippKopp** [comment]
Nice :) Thanks for letting me know and the explanations! Always good to get feedback on how to improve my code :)

**Comment by PhilippKopp** [comment]
Hi @Dobiasd 
Nice! 
I had a look at your changes to `prelu_layer.hpp`. Couldn't think of any cases that would fail. I would say it should have the exact same functionality as before.

**Comment by Dobiasd** [comment]
Cool, thanks a lot. :)
